test/libsolidity/ASTPropertyTest.cpp

```diff
@@ -0,0 +1,170 @@
+
+#include <boost/algorithm/string.hpp>
+#include <boost/throw_exception.hpp>
+
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/split.hpp>
+
+#include <queue>
+
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+using namespace solidity::frontend::test;
+using namespace solidity;
+using namespace std;
+
+ASTPropertyTest::ASTPropertyTest(string const& _filename):
+	TestCase(_filename)
+{
+	if (!boost::algorithm::ends_with(_filename, ".sol"))
+		BOOST_THROW_EXCEPTION(runtime_error("Invalid test contract file name: \"" + _filename + "\"."));
+
+	m_source = m_reader.source();
+	readExpectations();
+}
+
+void ASTPropertyTest::generateTestCaseValues(string& _values, bool _obtained)
+{
+	_values.clear();
+	for (auto testId: m_expectationsSequence)
+	{
+		soltestAssert(m_testCases.count(testId) > 0);
+		_values +=
+			testId +
+			": " +
+			(_obtained ? m_testCases[testId].obtainedValue : m_testCases[testId].expectedValue)
+			+ "\n";
+	}
+}
+
+vector<StringPair> ASTPropertyTest::readKeyValuePairs(string const& _input)
+{
+	vector<StringPair> result;
+	for (string const& line: _input | ranges::views::split('\n') | ranges::to<vector<string>>)
+	{
+		if (line.empty())
+			continue;
+
+		vector<string> pair;
+		for (string const& element: line | ranges::views::split(':') | ranges::to<vector<string>>)
+		{
+			soltestAssert(ranges::all_of(boost::trim_copy(element), [](char c) { return isprint(c); }));
+			pair.emplace_back(boost::trim_copy(element));
+		}
+
+		soltestAssert(pair.size() == 2);
+		result.emplace_back(pair[0], pair[1]);
+	}
+	return result;
+}
+
+void ASTPropertyTest::readExpectations()
+{
+	for (auto const& [testId, testExpectation]: readKeyValuePairs(m_reader.simpleExpectations()))
+	{
+		m_testCases.emplace(testId, ASTPropertyTestCase{testId, "", testExpectation, ""});
+		m_expectationsSequence.push_back(testId);
+	}
+	generateTestCaseValues(m_expectation, false);
+}
+
+optional<Json::Value> ASTPropertyTest::findNode(Json::Value const& _root, string_view const& _property)
+{
+	if (!_property.empty())
+	{
+		string subNode = string(_property.substr(0, _property.find_first_of('.')));
+		if (subNode != _property)
+			return findNode(_root[subNode], _property.substr(_property.find_first_of('.') + 1));
+		else if (_root.isMember(subNode))
+			return make_optional(_root[subNode]);
+	}
+	return {};
+}
+
+void ASTPropertyTest::readTestedProperties(Json::Value const& _astJson)
+{
+	queue<Json::Value> nodesToVisit;
+	nodesToVisit.push(_astJson);
+	string documentation = "documentation";
+	string testCaseLine;
+
+	while(!nodesToVisit.empty())
+	{
+		auto& node = nodesToVisit.front();
+		if (node.isObject())
+		{
+			for (auto memberName: node.getMemberNames())
+				if (memberName == documentation)
+				{
+					auto docNode = node[documentation];
+					testCaseLine = docNode.isObject() ?
+						docNode["text"].asString() :
+						docNode.asString();
+
+					soltestAssert(!testCaseLine.empty());
```

Comment:
If you put more then one property test on a node, this code silently ignores that.

It probably also lets tests with duplicate IDs overwrite each other.